File: belaberung-js-shared/src/messages.ts
import { Room } from "./rooms.js";
import { User } from "./users.js";

export class Message {
  id: number = 0;
  content: string;
  attachment: MessageAttachment;
  timestamp: Date;
  user: User;
  room: Room;
  constructor(
    content: string,
    attachment: MessageAttachment,
    user: User,
    room: Room,
  ) {
    this.content = content;
    this.attachment = attachment;
    this.user = user;
    this.room = room;
    this.timestamp = new Date();
  }
}

export enum MessageAttachmentType {
  None = "None",
  Mention = "Mention",
  Reply = "Reply",
  BinaryFile = "BinaryFile",
  TextFile = "TextFile",
  Document = "Document",
  Hyperlink = "Hyperlink",
  Image = "Image",
  Audio = "Audio",
  Video = "Video",
}

export class MessageAttachment {
  type: MessageAttachmentType;
  data: any;
  constructor(type: MessageAttachmentType, data: any) {
    this.type = type;
    this.data = data;
  }
}

File: belaberung-go/model/messages.go
package model

import (
	"context"
	"encoding/json"
	"time"

	"github.com/uptrace/bun"
)

type MessageAttachmentType string

const (
	MessageAttachmentNone       MessageAttachmentType = "None"
	MessageAttachmentMention    MessageAttachmentType = "Mention"
	MessageAttachmentReply      MessageAttachmentType = "Reply"
	MessageAttachmentBinaryFile MessageAttachmentType = "BinaryFile"
	MessageAttachmentTextFile   MessageAttachmentType = "TextFile"
	MessageAttachmentDocument   MessageAttachmentType = "Document"
	MessageAttachmentHyperlink  MessageAttachmentType = "Hyperlink"
	MessageAttachmentImage      MessageAttachmentType = "Image"
	MessageAttachmentAudio      MessageAttachmentType = "Audio"
	MessageAttachmentVideo      MessageAttachmentType = "Video"
)

type MessageAttachment struct {
	Type MessageAttachmentType `json:"type"`
	Data json.RawMessage       `json:"data"`
}

type Message struct {
	bun.BaseModel `bun:"table:messages"`

	ID int `bun:"id,pk,autoincrement"`

	Content string

	Attachment MessageAttachment `bun:",type:jsonb"`

	Timestamp time.Time `bun:",index:idx_room_time"`

	UserID int `bun:",notnull,index:idx_user_messages"`
	RoomID int `bun:",notnull,index:idx_room_time"`

	User *User `bun:"rel:belongs-to,join:user_id=id"`
	Room *Room `bun:"rel:belongs-to,join:room_id=id"`
}

func CreateMessage(ctx context.Context, db *bun.DB, content string, attachment MessageAttachment, userID, roomID int) (*Message, error) {
	message := &Message{
		Content:    content,
		Attachment: attachment,
		Timestamp:  time.Now(),
		UserID:     userID,
		RoomID:     roomID,
	}

	_, err := db.NewInsert().
		Model(message).
		Exec(ctx)
	if err != nil {
		return nil, err
	}

	return message, nil
}
